### PR TITLE
[fix bug 1144183] People end-point for the KPI diagram.

### PR DESCRIPTION
**Pull request by akatsoulas**

### CODE REVIEW

File: remo/profiles/api/views.py [hunk 1]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+            start = current_week_start - timedelta(weeks=i)
+            end = start + timedelta(weeks=1)
+
+            # Total number of reps (per week) for the past 6 weeks
+            count = people.qs.filter(
+                date_joined_program__range=[start, end]).count()
+            weekly_count.append({'week': weeks-i, 'people': count})
+
+        # Query params
+        q_kwargs = {'user__ng_reports__report_date__range': None}
```
Comment (johngian) [inline remo/profiles/api/views.py]: Please use ``Q()`` objects instead of ``**q_kwargs``.

File: remo/profiles/api/views.py [hunk 2]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+
+        # Activity metrics:
+        # Inactive: No activity within 16 weeks (8 past, 8 future)
+        # Casual: 1 activity within 16 weeks (8 past, 8 future)
+        # Active: 1 activity within 8 weeks (4 past, 4 future)
+        # Core: 4 activities within 8 weeks (4 past, 4 future)
+        past_four_weeks = today - timedelta(weeks=4)
+        future_four_weeks = today + timedelta(weeks=4)
+        eight_weeks_range = [past_four_weeks, future_four_weeks]
+        q_kwargs['user__ng_reports__report_date__range'] = eight_weeks_range
```
Comment (johngian) [inline remo/profiles/api/views.py]: Same here.

File: remo/profiles/api/views.py [hunk 3]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+                                                       num_reports__lt=CORE)
+        num_active = active_contributors.count()
+        # Core contributors
+        num_core = core_active_query.filter(num_reports__gte=CORE).count()
+
+        # Query for sixteen weeks range
+        past_eight_weeks = today - timedelta(weeks=MAX_ACTIVITY_PERIOD)
+        future_eight_weeks = today + timedelta(weeks=MAX_ACTIVITY_PERIOD)
+        sixteen_weeks_range = [past_eight_weeks, future_eight_weeks]
+        q_kwargs['user__ng_reports__report_date__range'] = sixteen_weeks_range
```
Comment (johngian) [inline remo/profiles/api/views.py]: Same here.

File: remo/profiles/api/views.py [hunk 4]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+            past_eight_weeks = start - timedelta(weeks=MAX_ACTIVITY_PERIOD)
+            future_eight_weeks = start + timedelta(weeks=MAX_ACTIVITY_PERIOD)
+            past_four_weeks = start - timedelta(weeks=4)
+            future_four_weeks = start + timedelta(weeks=4)
+
+            # Converion points per week
+            eight_weeks = [past_four_weeks, future_four_weeks]
+            sixteen_weeks = [past_eight_weeks, future_eight_weeks]
+
+            q_kwargs['user__ng_reports__report_date__range'] = eight_weeks
```
Comment (johngian) [inline remo/profiles/api/views.py]: Same here.

File: remo/profiles/api/views.py [hunk 5]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+                                     num_reports=Count('user__ng_reports')))
+            # Active contributors
+            active_contributors = core_active_query.filter(
+                num_reports__gte=ACTIVE, num_reports__lt=CORE)
+            active_weekly = active_contributors.count()
+            # Core contributors
+            core_weekly = core_active_query.filter(
+                num_reports__gte=CORE).count()
+
+            q_kwargs['user__ng_reports__report_date__range'] = sixteen_weeks
```
Comment (johngian) [inline remo/profiles/api/views.py]: Same here.

File: remo/profiles/api/views.py [hunk 6]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+            # Total number of reps (per week) for the past 6 weeks
+            count = people.qs.filter(
+                date_joined_program__range=[start, end]).count()
+            weekly_count.append({'week': weeks-i, 'people': count})
+
+        # Query params
+        q_kwargs = {'user__ng_reports__report_date__range': None}
+        # Get the number of reports for each user.
+
+        # Activity metrics:
```
Comment (johngian) [inline remo/profiles/api/views.py]: I suggest you implement some helper methods to get ``inactive``, ``casual``, ``active``, ``core`` for a specific date.

**Comment by johngian** [comment]
You could deduplicate some code under ``per_week`` calculations which has the same logic with the calculation for ``this_week``.

File: remo/profiles/api/views.py (hunk 1 above)
Comment (akatsoulas) [inline remo/profiles/api/views.py]: I am not sure this is the best case scenario to use `Q` objects. The only reason `q_kwargs` is used here is to make the code more readable. I don't expect `Q()` object to help that much. Thoughts?

File: remo/profiles/api/views.py [hunk 7]
```diff
@@ -24,3 +45,179 @@ def retrieve(self, request, pk):
...
+        weekly_contribution = []
+        for i in range(weeks):
+            start = current_week_start - timedelta(weeks=i)
+
+            past_eight_weeks = start - timedelta(weeks=MAX_ACTIVITY_PERIOD)
+            future_eight_weeks = start + timedelta(weeks=MAX_ACTIVITY_PERIOD)
+            past_four_weeks = start - timedelta(weeks=4)
+            future_four_weeks = start + timedelta(weeks=4)
+
+            # Converion points per week
```
Comment (johngian) [inline remo/profiles/api/views.py]: Please fix typo

File: remo/profiles/api/views.py (hunk 1 above)
Comment (johngian) [inline remo/profiles/api/views.py]: * ``Q()`` objects purpose is to encapsulate a collection of keyword arguments, so for us it make sense
* Using ``Q()`` object looks more clean and readable to me
* I don't really like the idea of using ``**kwargs`` when django provides ``Q()``

**Comment by johngian** [comment]
Please split this in 2 commits. One with the People KPI endpoint and the other one with the rest of the changes.

**Comment by johngian** [comment]
Please add some tests like we do in the rest of the kpi endpoints.

**Comment by johngian** [comment]
R+ bravisimo

**Pull request merged**